trieste/models/keras/sampler.py

```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
+        """
+        Generate an approximate function draw (trajectory) by using the predicted means
+    of the stochastic forward passes as a trajectory.
+
+        :return: A trajectory function representing an approximate trajectory
+            from the model, taking an input of shape `[N, B, D]` and returning shape `[N, B]`. 
+            N is the number of samples in each batch and B is the batch size. 
+        """
+
+        return deep_evidential_trajectory(self._model)
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently resample a :const:`TrajectoryFunction` in-place to avoid function retracing
+        with every new sample.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, deep_evidential_trajectory))
+        trajectory.resample()
+        return trajectory
+
+class deep_evidential_trajectory(TrajectoryFunctionClass):
+    """
+    Generate an approximate function draw (trajectory) by drawing a mean and variance
+    vector from the posterior distributions.
+    """
+
+    def __init__(self, model: EvidentialPriorModel):
+        """
+        :param model: The model of the objective function.
+        """
+        self._model = model
+        self._initialized = tf.Variable(False, trainable=False)
+
+        self._batch_size = tf.Variable(
+            0, dtype=tf.int32, trainable=False
+        )
+        self._evidential_parameters = None
+        self._mu = tf.Variable(0, trainable=False)
+        self._sigma = tf.Variable(0, trainable=False)
+            
+    def __call__(self, x: TensorType) -> TensorType:  # [N, B, d] -> [N, B]
+        """
+        Call trajectory function. Makes a draw from the posterior normal distribution
+        of outputs given the predicted evidential parameters. 
+        """
+
+        tf.debugging.assert_equal(
+            self._model.model.flattened_output_shape,
+            tf.shape(x)[-1],
+            message=f"""
+            IS THIS EVEN THE RIGHT TEST??
+            """
+         )
+
+        if not self._initialized:  # work out desired batch size from input
+            self._batch_size.assign(tf.shape(x)[-2])  # B
+            self._evidential_parameters = self._model.model(x)
+            self.resample() # Draws a mu and sigma vector
+            self._initialized.assign(True)
+
+        tf.debugging.assert_equal(
+            tf.shape(x)[-2],
+            self._batch_size.value(),
+            message=f"""
+            This trajectory only supports batch sizes of {self._batch_size}.
+            If you wish to change the batch size you must get a new trajectory
+            by calling the get_trajectory method of the trajectory sampler.
+            """
+        )
+
+        dist = tfp.distributions.Normal(self._mu, self._sigma)
```

Comment:
I pretty much just stopped working on this as soon as I realized I didn't understand the way the dimensions were working and waited for Hrvoje's response to my questions. Haven't touched the sampler since. Still a bit confusing to me at first thought whether it is the prediction which would just be mu or the distribution that constitutes the trajectory

Reply:
In the case of ensembles and you can do either (notice he has this "use_samples" argument to return the distribution and sample from it, and otherwise the default behavior is the predictions from networks in the ensemble). Since I'm also stuck at the sampler, let's just see what Hrvoje has to say, and (if you don't mind) I'll take care of it.